src/server/services/wfs/qgswfsdescribefeaturetype.cpp

```diff
@@ -133,281 +93,103 @@ namespace QgsWfs
+      typeNameList = wfsParams.typeNames();
     }
 
-    const QStringList wfsLayerIds = QgsServerProjectUtils::wfsLayerIds( *project );
-    for ( int i = 0; i < wfsLayerIds.size(); ++i )
-    {
-      QgsMapLayer *layer = project->mapLayer( wfsLayerIds.at( i ) );
-      if ( !layer )
-      {
-        continue;
-      }
-
-      const QString name = layerTypeName( layer );
-
-      if ( !typeNameList.isEmpty() && !typeNameList.contains( name ) )
-      {
-        continue;
-      }
-#ifdef HAVE_SERVER_PYTHON_PLUGINS
-      if ( accessControl && !accessControl->layerReadPermission( layer ) )
-      {
-        if ( !typeNameList.isEmpty() )
-        {
-          throw QgsSecurityAccessException( QStringLiteral( "Feature access permission denied" ) );
-        }
-        else
-        {
-          continue;
-        }
-      }
-#endif
-      QgsVectorLayer *vLayer = qobject_cast<QgsVectorLayer *>( layer );
-      QgsVectorDataProvider *provider = vLayer->dataProvider();
-      if ( !provider )
-      {
-        continue;
-      }
-      setSchemaLayer( schemaElement, doc, const_cast<QgsVectorLayer *>( vLayer ), oFormat );
-    }
-    return doc;
+    return typeNameList;
   }
 
-  void setSchemaLayer( QDomElement &parentElement, QDomDocument &doc, const QgsVectorLayer *layer, QgsWfsParameters::Format format )
-  {
-    const QgsVectorDataProvider *provider = layer->dataProvider();
-    if ( !provider )
-    {
-      return;
-    }
-
-    const QString typeName = layerTypeName( layer );
 
-    //xsd:element
-    QDomElement elementElem = doc.createElement( QStringLiteral( "element" )/*xsd:element*/ );
-    elementElem.setAttribute( QStringLiteral( "name" ), typeName );
-    elementElem.setAttribute( QStringLiteral( "type" ), "qgs:" + typeName + "Type" );
-    elementElem.setAttribute( QStringLiteral( "substitutionGroup" ), QStringLiteral( "gml:_Feature" ) );
-    parentElement.appendChild( elementElem );
-
-    //xsd:complexType
-    QDomElement complexTypeElem = doc.createElement( QStringLiteral( "complexType" )/*xsd:complexType*/ );
-    complexTypeElem.setAttribute( QStringLiteral( "name" ), typeName + "Type" );
-    parentElement.appendChild( complexTypeElem );
-
-    //xsd:complexType
-    QDomElement complexContentElem = doc.createElement( QStringLiteral( "complexContent" )/*xsd:complexContent*/ );
-    complexTypeElem.appendChild( complexContentElem );
+  void getFieldAttributes( const QgsField &field, QString &fieldName, QString &fieldType )
+  {
+    fieldName = field.name();
 
-    //xsd:extension
-    QDomElement extensionElem = doc.createElement( QStringLiteral( "extension" )/*xsd:extension*/ );
-    extensionElem.setAttribute( QStringLiteral( "base" ), QStringLiteral( "gml:AbstractFeatureType" ) );
-    complexContentElem.appendChild( extensionElem );
+    const thread_local QRegularExpression sCleanTagNameRegExp( QStringLiteral( "[^\\w\\.-_]" ), QRegularExpression::PatternOption::UseUnicodePropertiesOption );
```

Comment:
Just wondering if this tag-cleaning is done elsewhere, if that's the case it would be nice to have it in a separate static method.

Reply:
Moved in static method in `qgswfsutils.cpp`. Found two more occurences in `qgswfsgetfeature.cpp`